### PR TITLE
[RLlib] Policy-classes cleanup and torch/tf unification.

### PR DESCRIPTION
**Pull request by sven1977**
## Why are these changes needed?

- Make Policy abstract.
- Add action_dist to call to extra_action_out_fn (necessary for PPO torch).
- Move some methods and vars to base Policy (from TFPolicy): num_state_tensors, ACTION_PROB, ACTION_LOGP and some more.
- General cleanup and getting RLlib ready for torch PPO.

Issue #3365

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by ericl** [comment]
Is this ready for review? It's marked as a draft.

**Comment by sven1977** [comment]
Just waiting for the tests to pass, now that we fixed the DQN-Keras thing and the tf==2.0 issue.

**Comment by sven1977** [comment]
alpha zero had a hiccup w/ the new c'tor. New test run ...

File: rllib/agents/qmix/qmix_policy.py [hunk 1]
```diff
@@ -143,6 +143,7 @@ def forward(self,
         return loss, mask, masked_td_error, chosen_action_qvals, targets
 
 
+# TODO: Make this a TorchPolicy child.
```
Comment (robertnishihara) [inline rllib/agents/qmix/qmix_policy.py]: Can you use `TODO(sven)` or something like that? The point of this is so that people looking at the TODO in the future know who wrote it and can talk to them.

File: rllib/contrib/maddpg/maddpg_policy.py [hunk 2]
```diff
@@ -313,11 +308,10 @@ def _build_critic_network(self,
                               act_n,
                               obs_space_n,
                               act_space_n,
-                              hiddens,
-                              activation=None,
```
Comment (ericl) [inline rllib/contrib/maddpg/maddpg_policy.py]: Why this change? It makes the parameters less explicit

File: rllib/evaluation/sampler.py [hunk 3]
```diff
@@ -587,7 +587,7 @@ def _process_policy_eval_results(to_eval, eval_results, active_episodes,
             env_id = eval_data[i].env_id
             agent_id = eval_data[i].agent_id
             if clip_actions:
```
Comment (ericl) [inline rllib/evaluation/sampler.py]: It's better to use bare functions instead of a class if possible (principle of least power)

File: rllib/policy/policy.py [hunk 4]
```diff
@@ -211,7 +221,7 @@ def get_weights(self):
         Returns:
             weights (obj): Serializable copy or view of model weights
         """
-        raise NotImplementedError
+        pass
```
Comment (ericl) [inline rllib/policy/policy.py]: Why make this change?

File: rllib/policy/policy.py [hunk 5]
```diff
@@ -272,27 +290,29 @@ def export_checkpoint(self, export_dir):
         """
         raise NotImplementedError
 
+    @staticmethod
+    def clip_action(action, space):
```
Comment (ericl) [inline rllib/policy/policy.py]: This doesn't need to be a class member

**Review by ericl — CHANGES_REQUESTED** [review]
Some comments

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py]: I don't want to make this an `@abstractmethod`, b/c then we have to implement it (using probably just `pass` anyway) for all our "dummy" Policies that we use for testing.

File: rllib/policy/policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/policy.py]: Will change back, but the `clip_actions` belongs either into utils. Or: If only Policies ever use it, then yes, it should be a staticmethod.

File: rllib/agents/qmix/qmix_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/qmix/qmix_policy.py]: Sure.

File: rllib/contrib/maddpg/maddpg_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: Yes, it's a simplification. If `hiddens` and `activation` are information that is already given within the config dict (which is also passed into this function), then why pass them in as separate params on top? Also, it's a private helper function that's only called within this class 3 times.

Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: Ah, I remember now. The problem is `self.config`, which should not be defined at this point as config is now a super-class ctor argument (for unification reasons: all Policies actors have this arg now) and the super constructor has not been called in this call to _build_critic_network().

File: rllib/evaluation/sampler.py (hunk 3 above)
Comment (sven1977) [reply rllib/evaluation/sampler.py]: Ok, it is stateless. We should move it to utils, though.

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py]: Like: TestPolicy, StatefulPolicy, MockPolicy, etc..

File: rllib/contrib/maddpg/maddpg_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: If we changed it back, we would consequently have to pass in `use_state_preprocessor` as an arg as well. The question would be: Why is it ok to pull some information from self.config, but others has to be passed into this function explicitly? It's not consistent.

Comment (ericl) [reply rllib/contrib/maddpg/maddpg_policy.py]: One feedback I've gotten from users before is that reading directly from config makes it hard to understand what are the parameters of functions (i.e., it's fine to have a config at the top level, but confusing if it's passed whole to lower level modules). It's probably true we pass config dicts around more than necessary and would be better off with python named args / kwargs instead.

File: rllib/policy/policy.py (hunk 4 above)
Comment (ericl) [reply rllib/policy/policy.py]: ah, I mean why not keep it as a raise?

File: rllib/evaluation/sampler.py (hunk 3 above)
Comment (sven1977) [reply rllib/evaluation/sampler.py]: Moved back.

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py]: Then the new ABCMeta metaclass of the base Policy would require this method to be implemented (raise or not) in all the above dummies.

File: rllib/policy/policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/policy.py]: Moved back to being a function.

File: rllib/agents/qmix/qmix_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/qmix/qmix_policy.py]: Done.

File: rllib/contrib/maddpg/maddpg_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: Ok, makes sense. I'll pass in all three as single params, no `config` anymore. ...

Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: Done.

File: rllib/contrib/maddpg/maddpg_policy.py [hunk 6]
```diff
@@ -130,8 +130,9 @@ def _make_ph_n(space_n, name=""):
             new_act_ph_n,
             obs_space_n,
             act_space_n,
-            hiddens=config["critic_hiddens"],
-            activation=getattr(tf.nn, config["critic_hidden_activation"]),
+            config["use_state_preprocessor"],
```
Comment (ericl) [inline rllib/contrib/maddpg/maddpg_policy.py]: In the future, prefer to pass named args with the name for readability.

**Review by ericl — APPROVED** [review]

File: rllib/contrib/maddpg/maddpg_policy.py (hunk 6 above)
Comment (sven1977) [reply rllib/contrib/maddpg/maddpg_policy.py]: Yes, and we should start using *s in the method signature (only py3 supports those) to enforce using named args after maybe the initial 1-2 first arguments (which are often a given and maybe don't need to be named).

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py]: Actually, you are right, we could keep the `raise` here and still have it a non `@abstractmethod`. I'll change it back in a future PR.
